Cython/Utility/ModuleSetupCode.c

```
@@ -2231,6 +2231,105 @@ static PyObject* __Pyx_PyCode_New(
+  //  of PyObject.  PyRun_SimpleStringFlags and PySys_GetObject both look fairly safe to
+  //  use like this.
+  //
+  //  Since PyRun_SimpleStringFlags runs in the __main__ scope, we should be careful not
+  //  to define variables inside it.
+  //  We use PyRun_SimpleStringFlags instead of PyRun_SimpleString since SimpleString is
+  //  a macro, and MSVC doesn't like #ifdef within a macro expansion.
+  //
+  //  PyRun_SimpleStringFlags prints and clears the exception information.  We raise
+  //  therefore raise the proper exception from C (I think the interface to PyErr_SetString
+  //  should be consistent enough between builds for this to be OK). This is fragile but
+  //  the best we can do.
+  //
+  #if CYTHON_COMPILING_IN_CPYTHON
+    if (PySys_GetObject("gettotalrefcount")) {
+      #ifndef Py_DEBUG
+        PyErr_SetString(
+            PyExc_ImportError,
+            "Module was compiled with a non-debug version of Python but imported into a debug version."
+        );
+        return -1;
+      #endif
+    } else {
+      #ifdef Py_DEBUG
+        PyErr_SetString(
+            PyExc_ImportError,
+            "Module was compiled with a debug version of Python but imported into a non-debug version."
+        );
+        return -1;
+      #endif
+    }
+    #if PY_VERSION_HEX >= 0x030d0000
+    if (PyRun_SimpleStringFlags(
+      "if "
+      #ifdef Py_GIL_DISABLED
+        "not "
+      #endif
+      "__import__('sysconfig').get_config_var('Py_GIL_DISABLED'):\n"
+      "  raise ImportError",
+      NULL
+    ) == -1) {
+        PyErr_SetString(
+            PyExc_ImportError,
+      #ifdef Py_GIL_DISABLED
+            "Module was compiled with a freethreading build of Python but imported into a non-freethreading build."
+      #else
+            "Module was compiled with a non-freethreading build of Python but imported into a freethreading build."
+      #endif
+        );
```

You could make the exception message a part of the Python code (i.e. `raise ImportError('message')` to avoid raising the exception twice.

---

That was my initial thought but unfortunately it doesn't work. `PyRun_SimpleStringFlags` clears the exception and there's no way to get it back. The interfaces that don't swallow the exception mostly require creating Python objects, which I'm trying to avoid if we're linked to the wrong ABI.